// coinbase-pro-node/dist/account/AccountAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountAPI = exports.AccountType = void 0;
var AccountType;
(function (AccountType) {
    AccountType["FIAT"] = "fiat";
    AccountType["WALLET"] = "wallet";
})(AccountType = exports.AccountType || (exports.AccountType = {}));
class AccountAPI {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    /**
     * Get information for a single account. API key must belong to the same profile as the account.
     *
     * @param accountId - Account ID belonging to the API key’s profile
     * @see https://docs.pro.coinbase.com/#get-an-account
     */
    getAccount(accountId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${AccountAPI.URL.ACCOUNTS}/${accountId}`;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
    /**
     * List account activity of the API key’s profile. Account activity either increases or decreases your account
     * balance. Items are paginated and sorted latest first.
     *
     * @param accountId - Account ID belonging to the API key’s profile
     * @param pagination - Pagination field
     * @see https://docs.pro.coinbase.com/#get-account-history
     */
    getAccountHistory(accountId, pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${AccountAPI.URL.ACCOUNTS}/${accountId}/ledger`;
            const response = yield this.apiClient.get(resource, { params: pagination });
            return {
                data: response.data,
                pagination: {
                    after: response.headers['cb-after'],
                    before: response.headers['cb-before'],
                },
            };
        });
    }
    /**
     * List holds of an account that belong to the same profile as the API key. Holds are placed on an account for any
     * active orders or pending withdraw requests. As an order is filled, the hold amount is updated. If an order is
     * canceled, any remaining hold is removed. For a withdraw, once it is completed, the hold is removed.
     *
     * @param accountId - Account ID belonging to the API key’s profile
     * @param pagination - Pagination field
     * @see https://docs.pro.coinbase.com/#get-holds
     */
    getHolds(accountId, pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${AccountAPI.URL.ACCOUNTS}/${accountId}/holds`;
            const response = yield this.apiClient.get(resource, { params: pagination });
            return {
                data: response.data,
                pagination: {
                    after: response.headers['cb-after'],
                    before: response.headers['cb-before'],
                },
            };
        });
    }
    /**
     * Get a list of trading accounts from the profile of the API key.
     *
     * @see https://docs.pro.coinbase.com/#list-accounts
     */
    listAccounts() {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = AccountAPI.URL.ACCOUNTS;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
    /**
     * Get a list of your coinbase accounts.
     *
     * @see https://docs.pro.coinbase.com/#coinbase-accounts
     */
    listCoinbaseAccounts() {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = AccountAPI.URL.COINBASE_ACCOUNT;
            const response = yield this.apiClient.get(resource);
            return response.data;
        });
    }
}
exports.AccountAPI = AccountAPI;
AccountAPI.URL = {
    ACCOUNTS: `/accounts`,
    COINBASE_ACCOUNT: `/coinbase-accounts`,
};
