// coinbase-pro-node/dist/CoinbasePro.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoinbasePro = void 0;
const RESTClient_1 = require("./client/RESTClient");
const WebSocketClient_1 = require("./client/WebSocketClient");
const RequestSigner_1 = require("./auth/RequestSigner");
class CoinbasePro {
    constructor(auth = {
        apiKey: '',
        apiSecret: '',
        passphrase: '',
        useSandbox: false,
    }) {
        this.clockSkew = -1;
        if (typeof auth.useSandbox === 'boolean') {
            this.url =
                auth.useSandbox === true
                    ? CoinbasePro.SETUP.SANDBOX
                    : CoinbasePro.SETUP.PRODUCTION;
        }
        else {
            const { httpUrl, wsUrl } = auth;
            this.url = {
                REST: httpUrl,
                WebSocket: wsUrl,
            };
        }
        const signRequest = (setup) => __awaiter(this, void 0, void 0, function* () {
            /**
             * Cache clock skew to reduce the amount of API requests:
             * https://github.com/bennycode/coinbase-pro-node/issues/368#issuecomment-762122575
             */
            if (this.clockSkew === -1) {
                const time = yield this.rest.time.getTime();
                this.clockSkew = yield this.rest.time.getClockSkew(time);
            }
            return RequestSigner_1.RequestSigner.signRequest(auth, setup, this.clockSkew);
        });
        this.rest = new RESTClient_1.RESTClient(this.url.REST, signRequest);
        this.ws = new WebSocketClient_1.WebSocketClient(this.url.WebSocket, signRequest);
    }
}
exports.CoinbasePro = CoinbasePro;
CoinbasePro.SETUP = {
    PRODUCTION: {
        REST: 'https://api.pro.coinbase.com',
        WebSocket: 'wss://ws-feed.pro.coinbase.com',
    },
    SANDBOX: {
        REST: 'https://api-public.sandbox.pro.coinbase.com',
        WebSocket: 'wss://ws-feed-public.sandbox.pro.coinbase.com',
    },
};
