// coinbase-pro-node/dist/product/CandleBucketUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CandleBucketUtil = void 0;
/** The maximum number of data points for a single historic rates API request on Coinbase Pro is 300 candles: https://docs.pro.coinbase.com/#get-historic-rates */
const MAXIMUM_HISTORIC_DATA_POINTS = 300;
class CandleBucketUtil {
    static getMinPrice(candles, property = 'close') {
        const values = candles.map(candle => candle[property]);
        return Math.min(...values);
    }
    static getMaxPrice(candles, property = 'close') {
        const values = candles.map(candle => candle[property]);
        return Math.max(...values);
    }
    static addUnitMillis(openTime, granularityInSeconds, amount) {
        const granularityInMillis = granularityInSeconds * 1000;
        const units = amount * granularityInMillis;
        return new Date(openTime).getTime() + units;
    }
    static addUnitISO(openTime, granularityInSeconds, amount) {
        const nextTimestamp = CandleBucketUtil.addUnitMillis(openTime, granularityInSeconds, amount);
        return new Date(nextTimestamp).toISOString();
    }
    static removeUnitMillis(openTime, granularityInSeconds, amount) {
        const granularityInMillis = granularityInSeconds * 1000;
        const units = amount * granularityInMillis;
        return new Date(openTime).getTime() - units;
    }
    static removeUnitISO(openTime, granularityInSeconds, amount) {
        const nextTimestamp = CandleBucketUtil.removeUnitMillis(openTime, granularityInSeconds, amount);
        return new Date(nextTimestamp).toISOString();
    }
    static getIntervals() {
        return [60, 300, 900, 3600, 21600, 86400];
    }
    static mapInterval(intervals, interval) {
        return intervals.reduce((previous, current) => {
            return Math.abs(current - interval) < Math.abs(previous - interval) ? current : previous;
        });
    }
    static mapGranularity(candleSizeInMillis) {
        return this.mapInterval(CandleBucketUtil.getIntervals(), candleSizeInMillis);
    }
    static expectedBuckets(fromInMillis, toInMillis, candleSizeInMillis) {
        const timeSpanInMillis = toInMillis - fromInMillis;
        return timeSpanInMillis / candleSizeInMillis;
    }
    static getBucketsInMillis(fromInMillis, toInMillis, candleSizeInMillis) {
        const bucketsInMillis = [];
        const batch = MAXIMUM_HISTORIC_DATA_POINTS * candleSizeInMillis;
        let current = fromInMillis;
        bucketsInMillis.push(current); // push initial start
        current = current + batch;
        while (current < toInMillis) {
            bucketsInMillis.push(current - 1); // intermediate stop
            bucketsInMillis.push(current); // intermediate start
            current = current + batch;
        }
        bucketsInMillis.push(toInMillis); // push initial stop
        return bucketsInMillis;
    }
    static getBucketsInISO(bucketsInMillis) {
        const bucketsInISO = [];
        for (let i = 0; i < bucketsInMillis.length - 1; i += 2) {
            const start = new Date(bucketsInMillis[i]).toISOString();
            const stop = new Date(bucketsInMillis[i + 1]).toISOString();
            bucketsInISO.push({
                start,
                stop,
            });
        }
        return bucketsInISO;
    }
}
exports.CandleBucketUtil = CandleBucketUtil;
