// coinbase-pro-node/dist/fill/FillAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FillAPI = exports.Liquidity = void 0;
var Liquidity;
(function (Liquidity) {
    Liquidity["MAKER"] = "M";
    Liquidity["TAKER"] = "T";
})(Liquidity = exports.Liquidity || (exports.Liquidity = {}));
class FillAPI {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    /**
     * Get a list of recent fills for a given Order of the API key's profile.
     *
     * @param orderId - ID of previously placed order
     * @param pagination - Pagination field
     * @see https://docs.pro.coinbase.com/#list-fills
     * @see https://pro.coinbase.com/orders/filled
     */
    getFillsByOrderId(orderId, pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = FillAPI.URL.FILLS;
            const response = yield this.apiClient.get(resource, { params: Object.assign({ order_id: orderId }, pagination) });
            return {
                data: response.data,
                pagination: {
                    after: response.headers['cb-after'],
                    before: response.headers['cb-before'],
                },
            };
        });
    }
    /**
     * Get a list of recent fills for a given Product of the API key's profile.
     *
     * @param productId - Representation for base and counter
     * @param pagination - Pagination field
     * @see https://docs.pro.coinbase.com/#list-fills
     * @see https://pro.coinbase.com/orders/filled
     */
    getFillsByProductId(productId, pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = FillAPI.URL.FILLS;
            const response = yield this.apiClient.get(resource, { params: Object.assign({ product_id: productId }, pagination) });
            return {
                data: response.data,
                pagination: {
                    after: response.headers['cb-after'],
                    before: response.headers['cb-before'],
                },
            };
        });
    }
}
exports.FillAPI = FillAPI;
FillAPI.URL = {
    FILLS: `/fills`,
};
