// coinbase-pro-node/dist/order/OrderAPI.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderAPI = exports.OrderStatus = exports.SelfTradePrevention = exports.CancelOrderPeriod = exports.TimeInForce = exports.OrderType = void 0;
const querystring_1 = __importDefault(require("querystring"));
var OrderType;
(function (OrderType) {
    OrderType["LIMIT"] = "limit";
    OrderType["MARKET"] = "market";
})(OrderType = exports.OrderType || (exports.OrderType = {}));
var TimeInForce;
(function (TimeInForce) {
    TimeInForce["FILL_OR_KILL"] = "FOK";
    TimeInForce["GOOD_TILL_CANCELED"] = "GTC";
    TimeInForce["GOOD_TILL_TIME"] = "GTT";
    TimeInForce["IMMEDIATE_OR_CANCEL"] = "IOC";
})(TimeInForce = exports.TimeInForce || (exports.TimeInForce = {}));
var CancelOrderPeriod;
(function (CancelOrderPeriod) {
    CancelOrderPeriod["ONE_DAY"] = "day";
    CancelOrderPeriod["ONE_HOUR"] = "hour";
    CancelOrderPeriod["ONE_MINUTE"] = "min";
})(CancelOrderPeriod = exports.CancelOrderPeriod || (exports.CancelOrderPeriod = {}));
// https://docs.pro.coinbase.com/#self-trade-prevention
var SelfTradePrevention;
(function (SelfTradePrevention) {
    SelfTradePrevention["CANCEL_BOTH"] = "cb";
    SelfTradePrevention["CANCEL_NEWEST"] = "cn";
    SelfTradePrevention["CANCEL_OLDEST"] = "co";
    SelfTradePrevention["DECREMENT_AND_CANCEL"] = "dc";
})(SelfTradePrevention = exports.SelfTradePrevention || (exports.SelfTradePrevention = {}));
var OrderStatus;
(function (OrderStatus) {
    OrderStatus["ACTIVE"] = "active";
    OrderStatus["DONE"] = "done";
    OrderStatus["OPEN"] = "open";
    OrderStatus["PENDING"] = "pending";
})(OrderStatus = exports.OrderStatus || (exports.OrderStatus = {}));
class OrderAPI {
    constructor(apiClient) {
        this.apiClient = apiClient;
    }
    /**
     * With best effort, cancel all open orders from the profile that the API key belongs to.
     *
     * @param productId - Representation for base and counter
     * @returns A list of ids of the canceled orders
     * @see https://docs.pro.coinbase.com/#cancel-all
     */
    cancelOpenOrders(productId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = OrderAPI.URL.ORDERS;
            const response = yield this.apiClient.delete(resource, {
                params: productId ? { product_id: productId } : {},
            });
            return response.data;
        });
    }
    /**
     * Cancel a previously placed order. Order must belong to the profile that the API key belongs to.
     *
     * @param orderId - ID of the order to cancel
     * @param productId - While not required, the request will be more performant if you include the product ID
     * @returns The ID of the canceled order
     * @see https://docs.pro.coinbase.com/#cancel-an-order
     */
    cancelOrder(orderId, productId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${OrderAPI.URL.ORDERS}/${orderId}`;
            const response = yield this.apiClient.delete(resource, {
                params: productId ? { product_id: productId } : {},
            });
            return response.data;
        });
    }
    /**
     * List your orders from the profile that the API key belongs to. Only open or un-settled orders are returned. As
     * soon as an order is no longer open and settled, it will no longer appear in the default request.
     *
     * @note Depending on your activity, fetching all data from this endpoint can take very long (measured already 25
     *   seconds!)
     * @param query - Available query parameters (Pagination, Product ID and/or Order Status)
     * @see https://docs.pro.coinbase.com/#list-orders
     */
    getOrders(query) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = OrderAPI.URL.ORDERS;
            const response = yield this.apiClient.get(`${resource}`, {
                params: query,
                paramsSerializer: querystring_1.default.stringify,
            });
            return {
                data: response.data,
                pagination: {
                    after: response.headers['cb-after'],
                    before: response.headers['cb-before'],
                },
            };
        });
    }
    /**
     * Get a single order by order id from the profile that the API key belongs to.
     *
     * @param orderId - ID of previously placed order
     * @see https://docs.pro.coinbase.com/#get-an-order
     */
    getOrder(orderId) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = `${OrderAPI.URL.ORDERS}/${orderId}`;
            try {
                const response = yield this.apiClient.get(resource);
                return response.data;
            }
            catch (error) {
                /**
                 * If the order is canceled the response may
                 * have status code 404 if the order had no matches.
                 */
                if (error.response.status === 404) {
                    return null;
                }
                throw error;
            }
        });
    }
    /**
     * You can place two types of orders: limit and market. Orders can only be placed if your account has sufficient
     * funds. Once an order is placed, your account funds will be put on hold for the duration of the order.
     *
     * @param newOrder - Order type and parameters
     * @see https://docs.pro.coinbase.com/#place-a-new-order
     */
    placeOrder(newOrder) {
        return __awaiter(this, void 0, void 0, function* () {
            const resource = OrderAPI.URL.ORDERS;
            const response = yield this.apiClient.post(resource, newOrder);
            return response.data;
        });
    }
}
exports.OrderAPI = OrderAPI;
OrderAPI.URL = {
    ORDERS: `/orders`,
};
