// coinbase-pro-node/dist/client/RESTClient.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RESTClient = void 0;
const axios_1 = __importDefault(require("axios"));
const account_1 = require("../account");
const order_1 = require("../order");
const product_1 = require("../product");
const user_1 = require("../user");
const fee_1 = require("../fee");
const fill_1 = require("../fill");
const querystring_1 = __importDefault(require("querystring"));
const profile_1 = require("../profile");
const axios_retry_1 = __importStar(require("axios-retry"));
const util_1 = __importDefault(require("util"));
const events_1 = require("events");
const ErrorUtil_1 = require("../error/ErrorUtil");
const currency_1 = require("../currency");
const withdraw_1 = require("../withdraw");
const transfer_1 = require("../transfer");
const time_1 = require("../time");
// eslint-disable-next-line no-redeclare
class RESTClient extends events_1.EventEmitter {
    constructor(baseURL, signRequest) {
        super();
        this.signRequest = signRequest;
        this.logger = util_1.default.debuglog('coinbase-pro-node');
        this.httpClient = axios_1.default.create({
            baseURL: baseURL,
            timeout: 50000,
        });
        axios_retry_1.default(this.httpClient, {
            retries: Infinity,
            retryCondition: (error) => {
                return axios_retry_1.isNetworkOrIdempotentRequestError(error) || ErrorUtil_1.inAirPlaneMode(error) || ErrorUtil_1.gotRateLimited(error);
            },
            retryDelay: (retryCount, error) => {
                const errorMessage = ErrorUtil_1.getErrorMessage(error);
                this.logger(`#${retryCount} There was an error querying "${error.config.baseURL}${error.config.url}": ${errorMessage}`);
                /**
                 * Rate limits:
                 * - 3 requests per second, up to 6 requests per second in bursts for public endpoints
                 * - 5 requests per second, up to 10 requests per second in bursts for private endpoints
                 * @see https://docs.pro.coinbase.com/#rate-limits
                 */
                return 1000;
            },
            shouldResetTimeout: true,
        });
        this.httpClient.interceptors.request.use((config) => __awaiter(this, void 0, void 0, function* () {
            const baseURL = String(config.baseURL);
            const url = String(config.url);
            const requestPath = url.replace(baseURL, '');
            const signedRequest = yield this.signRequest({
                httpMethod: String(config.method).toUpperCase(),
                payload: RESTClient.stringifyPayload(config),
                requestPath,
            });
            config.headers = Object.assign(Object.assign({}, config.headers), { 'CB-ACCESS-KEY': signedRequest.key, 'CB-ACCESS-PASSPHRASE': signedRequest.passphrase, 'CB-ACCESS-SIGN': signedRequest.signature, 'CB-ACCESS-TIMESTAMP': signedRequest.timestamp });
            return config;
        }));
        this.account = new account_1.AccountAPI(this.httpClient);
        this.currency = new currency_1.CurrencyAPI(this.httpClient);
        this.fee = new fee_1.FeeAPI(this.httpClient);
        this.fill = new fill_1.FillAPI(this.httpClient);
        this.order = new order_1.OrderAPI(this.httpClient);
        this.product = new product_1.ProductAPI(this.httpClient, this);
        this.profile = new profile_1.ProfileAPI(this.httpClient);
        this.time = new time_1.TimeAPI(baseURL);
        this.transfer = new transfer_1.TransferAPI(this.httpClient);
        this.user = new user_1.UserAPI(this.httpClient);
        this.withdraw = new withdraw_1.WithdrawAPI(this.httpClient);
    }
    get defaults() {
        return this.httpClient.defaults;
    }
    get interceptors() {
        return this.httpClient.interceptors;
    }
    static stringifyPayload(config) {
        if (config.data) {
            return JSON.stringify(config.data);
        }
        const params = querystring_1.default.stringify(config.params);
        return params ? `?${params}` : '';
    }
}
exports.RESTClient = RESTClient;
